Test Server Query for non-cached execution causes SDK state machines problem

## Actual Behavior
```
io.temporal.client.WorkflowQueryException: workflowId='f2454b12-fcb1-4b3e-9b29-2f02e33a6543', runId='c5e13f9c-0e3c-454f-adb7-e7dd5fea844d', workflowType='TestWorkflowWithQuery'}

	at io.temporal.internal.sync.WorkflowStubImpl.query(WorkflowStubImpl.java:331)
	at io.temporal.testing.TimeLockingInterceptor$TimeLockingWorkflowStub.query(TimeLockingInterceptor.java:172)
	at io.temporal.internal.sync.WorkflowInvocationHandler$SyncWorkflowInvocationHandler.queryWorkflow(WorkflowInvocationHandler.java:309)
	at io.temporal.internal.sync.WorkflowInvocationHandler$SyncWorkflowInvocationHandler.invoke(WorkflowInvocationHandler.java:272)
	at io.temporal.internal.sync.WorkflowInvocationHandler.invoke(WorkflowInvocationHandler.java:178)
	at com.sun.proxy.$Proxy18.query(Unknown Source)
	at io.temporal.workflow.queryTests.LocalActivityAndQueryInvestigationTest.testLocalActivityAndQuery(LocalActivityAndQueryInvestigationTest.java:68)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: io.temporal.internal.replay.InternalWorkflowTaskException: Failure handling event 5 of type 'EVENT_TYPE_MARKER_RECORDED' during execution. {PreviousStartedEventId=3, workflowTaskStartedEventId=3, Currently Processing StartedEventId=3}
	at io.temporal.internal.statemachines.WorkflowStateMachines.createEventProcessingException(WorkflowStateMachines.java:244)
	at io.temporal.internal.statemachines.WorkflowStateMachines.handleEventsBatch(WorkflowStateMachines.java:223)
	at io.temporal.internal.statemachines.WorkflowStateMachines.handleEvent(WorkflowStateMachines.java:197)
	at io.temporal.internal.replay.ReplayWorkflowRunTaskHandler.handleWorkflowTaskImpl(ReplayWorkflowRunTaskHandler.java:180)
	at io.temporal.internal.replay.ReplayWorkflowRunTaskHandler.handleQueryWorkflowTask(ReplayWorkflowRunTaskHandler.java:247)
	at io.temporal.internal.replay.ReplayWorkflowTaskHandler.handleWorkflowTaskWithQuery(ReplayWorkflowTaskHandler.java:118)
	at io.temporal.internal.replay.ReplayWorkflowTaskHandler.handleWorkflowTask(ReplayWorkflowTaskHandler.java:97)
	at io.temporal.internal.worker.WorkflowWorker$TaskHandlerImpl.handleTask(WorkflowWorker.java:277)
	at io.temporal.internal.worker.WorkflowWorker$TaskHandlerImpl.handle(WorkflowWorker.java:231)
	at io.temporal.internal.worker.WorkflowWorker$TaskHandlerImpl.handle(WorkflowWorker.java:173)
	at io.temporal.internal.worker.PollTaskExecutor.lambda$process$0(PollTaskExecutor.java:93)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalArgumentException: LocalActivity: invalid REQUEST_PREPARED->MARKER_RECORDED, transition history is [CREATED->CHECK_EXECUTION_STATE, EXECUTING->SCHEDULE]
	at io.temporal.internal.statemachines.StateMachine.executeTransition(StateMachine.java:140)
	at io.temporal.internal.statemachines.StateMachine.handleHistoryEvent(StateMachine.java:101)
	at io.temporal.internal.statemachines.EntityStateMachineBase.handleEvent(EntityStateMachineBase.java:67)
	at io.temporal.internal.statemachines.WorkflowStateMachines.handleLocalActivityMarker(WorkflowStateMachines.java:482)
	at io.temporal.internal.statemachines.WorkflowStateMachines.handleCommandEvent(WorkflowStateMachines.java:292)
	at io.temporal.internal.statemachines.WorkflowStateMachines.handleSingleEvent(WorkflowStateMachines.java:250)
	at io.temporal.internal.statemachines.WorkflowStateMachines.handleEventsBatch(WorkflowStateMachines.java:221)
	... 12 more

	at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:271)
	at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:252)
	at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:165)
	at io.temporal.api.workflowservice.v1.WorkflowServiceGrpc$WorkflowServiceBlockingStub.queryWorkflow(WorkflowServiceGrpc.java:3135)
	at io.temporal.internal.client.external.GenericWorkflowClientExternalImpl.lambda$query$5(GenericWorkflowClientExternalImpl.java:220)
	at io.temporal.internal.retryer.GrpcSyncRetryer.retry(GrpcSyncRetryer.java:61)
	at io.temporal.internal.retryer.GrpcRetryer.retryWithResult(GrpcRetryer.java:51)
	at io.temporal.internal.client.external.GenericWorkflowClientExternalImpl.query(GenericWorkflowClientExternalImpl.java:213)
	at io.temporal.internal.client.RootWorkflowClientInvoker.query(RootWorkflowClientInvoker.java:141)
	at io.temporal.internal.sync.WorkflowStubImpl.query(WorkflowStubImpl.java:324)
	... 18 more
```

## Root cause

The way Test Server forms a `PollWorkflowTaskResponse` for a legacy query during replay makes SDK state machines think that the last workflow task is actually getting executed for the first time instead of being replayed (`previousStartedEventId` and `startedEventId` have the same values as the last executed workflow task) causing a failure when SDK tries to actually match already existing events with commands. This issue appears specifically with the LocalActivity state machine.

--- a/build.gradle
+++ b/build.gradle
@@ -9,7 +9,7 @@ plugins {
     id 'org.cadixdev.licenser' version '0.6.1'
     id 'com.palantir.git-version' version "${palantirGitVersionVersion}" apply false
     id 'io.github.gradle-nexus.publish-plugin' version '1.1.0'
-    id 'com.diffplug.spotless' version '6.5.1' apply false
+    id 'com.diffplug.spotless' version '6.5.2' apply false
     id 'base'
 }
 

--- a/temporal-kotlin/build.gradle
+++ b/temporal-kotlin/build.gradle
@@ -9,7 +9,7 @@ plugins {
 //    id 'org.jetbrains.kotlin.jvm' version '1.5.32'
 //    id 'org.jetbrains.kotlin.jvm' version '1.6.20'
     id 'org.jetbrains.kotlin.jvm' version "${kotlinVersion}"
-    id 'org.jlleitschuh.gradle.ktlint' version '10.2.1'
+    id 'org.jlleitschuh.gradle.ktlint' version '10.3.0'
 }
 
 apply plugin: 'org.jetbrains.kotlin.jvm'

--- a/temporal-sdk/src/main/java/io/temporal/internal/common/WorkflowExecutionHistory.java
+++ b/temporal-sdk/src/main/java/io/temporal/internal/common/WorkflowExecutionHistory.java
@@ -31,7 +31,6 @@ import io.temporal.api.history.v1.History;
 import io.temporal.api.history.v1.HistoryEvent;
 import io.temporal.common.converter.DataConverterException;
 import java.util.List;
-import javax.annotation.Nullable;
 
 /**
  * Provides a wrapper with convenience methods over raw protobuf {@link History} object representing
@@ -123,7 +122,6 @@ public final class WorkflowExecutionHistory {
     return history.getEventsList();
   }
 
-  @Nullable
   public HistoryEvent getLastEvent() {
     int eventsCount = history.getEventsCount();
     return eventsCount > 0 ? history.getEvents(eventsCount - 1) : null;

--- a/temporal-sdk/src/main/java/io/temporal/internal/replay/ReplayWorkflowTaskHandler.java
+++ b/temporal-sdk/src/main/java/io/temporal/internal/replay/ReplayWorkflowTaskHandler.java
@@ -99,7 +99,7 @@ public final class ReplayWorkflowTaskHandler implements WorkflowTaskHandler {
 
   private Result handleWorkflowTaskWithQuery(
       PollWorkflowTaskQueueResponse.Builder workflowTask, Scope metricsScope) throws Exception {
-    boolean legacyQuery = workflowTask.hasQuery();
+    boolean directQuery = workflowTask.hasQuery();
     AtomicBoolean createdNew = new AtomicBoolean();
     WorkflowExecution execution = workflowTask.getWorkflowExecution();
     WorkflowRunTaskHandler workflowRunTaskHandler = null;
@@ -112,11 +112,17 @@ public final class ReplayWorkflowTaskHandler implements WorkflowTaskHandler {
       boolean finalCommand;
       Result result;
 
-      if (legacyQuery) {
+      if (directQuery) {
+        // Direct query happens when there is no reason (events) to produce a real persisted
+        // workflow task.
+        // But Server needs to notify the workflow about the query and get back the query result.
+        // Server creates a fake non-persisted a PollWorkflowTaskResponse with just the query.
+        // This WFT has no new events in the history to process
+        // and the worker response on such a WFT can't contain any new commands either.
         QueryResult queryResult =
             workflowRunTaskHandler.handleQueryWorkflowTask(workflowTask, workflowTask.getQuery());
         finalCommand = queryResult.isWorkflowMethodCompleted();
-        result = createLegacyQueryResult(workflowTask, queryResult, null);
+        result = createDirectQueryResult(workflowTask, queryResult, null);
       } else {
         // main code path, handle workflow task that can have an embedded query
         WorkflowTaskResult wftResult = workflowRunTaskHandler.handleWorkflowTask(workflowTask);
@@ -146,15 +152,15 @@ public final class ReplayWorkflowTaskHandler implements WorkflowTaskHandler {
 
       if (useCache) {
         cache.invalidate(execution, metricsScope, "Exception", e);
-        // If history if full and exception occurred then sticky session hasn't been established
-        // yet and we can avoid doing a reset.
+        // If history is full and exception occurred then sticky session hasn't been established
+        // yet, and we can avoid doing a reset.
         if (!isFullHistory(workflowTask)) {
           resetStickyTaskQueue(execution);
         }
       }
 
-      if (legacyQuery) {
-        return createLegacyQueryResult(workflowTask, null, e);
+      if (directQuery) {
+        return createDirectQueryResult(workflowTask, null, e);
       } else {
         // this call rethrows an exception in some scenarios
         return failureToWFTResult(workflowTask, e);
@@ -278,7 +284,7 @@ public final class ReplayWorkflowTaskHandler implements WorkflowTaskHandler {
         workflowType, null, failedRequest.build(), null, null, false);
   }
 
-  private Result createLegacyQueryResult(
+  private Result createDirectQueryResult(
       PollWorkflowTaskQueueResponseOrBuilder workflowTask, QueryResult queryResult, Throwable e) {
     RespondQueryTaskCompletedRequest.Builder queryCompletedRequest =
         RespondQueryTaskCompletedRequest.newBuilder()

--- a/temporal-sdk/src/main/java/io/temporal/internal/statemachines/WorkflowStateMachines.java
+++ b/temporal-sdk/src/main/java/io/temporal/internal/statemachines/WorkflowStateMachines.java
@@ -67,21 +67,36 @@ public final class WorkflowStateMachines {
   /**
    * EventId of the WorkflowTaskStarted event of the Workflow Task that was picked up by a worker
    * and triggered a current replay or execution. It's expected to be the last event in the history
-   * unless we are replaying from a JSON file.
+   * if we continue to execute the workflow.
+   *
+   * <p>For direct (legacy) queries, it may be:
+   *
+   * <ul>
+   *   <li>0 if it's query a closed workflow execution
+   *   <li>id of the last successfully completed Workflow Task if the workflow is not closed
+   * </ul>
+   *
+   * <p>Set from the "outside" from the PollWorkflowTaskQueueResponse. Not modified by the SDK state
+   * machines.
    */
   private long workflowTaskStartedEventId;
 
   /**
-   * EventId of the started event of the last successfully executed workflow task in the history.
+   * EventId of the WorkflowTaskStarted event of the last successfully executed Workflow Task in the
+   * history. This variable plays a critical role in state machines understanding if it replays code
+   * for an existing history, or it executes it first time and produces new commands.
+   *
+   * <p>Set from the "outside" of state machines from the PollWorkflowTaskQueueResponse. Not
+   * modified by the SDK state machines.
    */
   private long previousStartedEventId;
 
-  /** EventId of the last WorkflowTaskStartedEvent handled by these state machines. */
+  /** EventId of the last WorkflowTaskStarted event handled by these state machines. */
   private long currentStartedEventId;
 
   /**
    * EventId of the last event seen by these state machines. Events earlier than this one will be
-   * discarded
+   * discarded.
    */
   private long lastHandledEventId;
 
@@ -1038,7 +1053,7 @@ public final class WorkflowStateMachines {
 
   private String createShortCurrentStateMessagePostfix() {
     return String.format(
-        "{PreviousStartedEventId=%s, workflowTaskStartedEventId=%s, Currently Processing StartedEventId=%s}",
-        this.getLastStartedEventId(), this.workflowTaskStartedEventId, this.currentStartedEventId);
+        "{PreviousStartedEventId=%s, WorkflowTaskStartedEventId=%s, CurrentStartedEventId=%s}",
+        this.previousStartedEventId, this.workflowTaskStartedEventId, this.currentStartedEventId);
   }
 }

--- a/temporal-sdk/src/main/java/io/temporal/worker/WorkerFactory.java
+++ b/temporal-sdk/src/main/java/io/temporal/worker/WorkerFactory.java
@@ -352,6 +352,8 @@ public final class WorkerFactory {
     log.info("awaitTermination done: {}", this);
   }
 
+  // TODO we should hide an actual implementation of WorkerFactory under WorkerFactory interface and
+  // expose this method on the implementation only
   @VisibleForTesting
   WorkflowExecutorCache getCache() {
     return this.cache;

--- a/temporal-sdk/src/test/java/io/temporal/workflow/queryTests/LocalActivityAndQueryTest.java
+++ b/temporal-sdk/src/test/java/io/temporal/workflow/queryTests/LocalActivityAndQueryTest.java
@@ -17,7 +17,7 @@
  *  permissions and limitations under the License.
  */
 
-package io.temporal.workflow.activityTests;
+package io.temporal.workflow.queryTests;
 
 import static org.junit.Assert.assertEquals;
 
@@ -42,7 +42,6 @@ import java.util.Optional;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ForkJoinPool;
 import java.util.concurrent.ForkJoinTask;
-import org.junit.Assert;
 import org.junit.Rule;
 import org.junit.Test;
 
@@ -74,25 +73,19 @@ public class LocalActivityAndQueryTest {
 
     // Ensure that query doesn't see intermediate results of the local activities execution
     // as all these activities are executed in a single workflow task.
-    while (true) {
-      String queryResult = workflowStub.query();
-      Assert.assertTrue(queryResult, queryResult.equals("run4"));
-      List<ForkJoinTask<String>> tasks = new ArrayList<ForkJoinTask<String>>();
-      int threads = 30;
-      if (queryResult.equals("run4")) {
-        for (int i = 0; i < threads; i++) {
-          ForkJoinTask<String> task = ForkJoinPool.commonPool().submit(() -> workflowStub.query());
-          tasks.add(task);
-        }
-        for (int i = 0; i < threads; i++) {
-          assertEquals("run4", tasks.get(i).get());
-        }
-        break;
-      }
+    // Query should only see the result after executing all local activities.
+    List<ForkJoinTask<String>> tasks = new ArrayList<>();
+    int threads = 30;
+    for (int i = 0; i < threads; i++) {
+      ForkJoinTask<String> task = ForkJoinPool.commonPool().submit(workflowStub::query);
+      tasks.add(task);
+      Thread.sleep(5);
     }
-    String result = WorkflowStub.fromTyped(workflowStub).getResult(String.class);
-    assertEquals("done", result);
-    assertEquals("run4", workflowStub.query());
+    for (int i = 0; i < threads; i++) {
+      assertEquals("run4", tasks.get(i).get());
+    }
+
+    assertEquals("done", WorkflowStub.fromTyped(workflowStub).getResult(String.class));
     activitiesImpl.assertInvocations(
         "sleepActivity", "sleepActivity", "sleepActivity", "sleepActivity", "sleepActivity");
     HistoryEvent marker =

--- a/temporal-sdk/src/test/java/io/temporal/workflow/queryTests/QueryCausingReplayWithLocalActivityInLastWFTTest.java
+++ b/temporal-sdk/src/test/java/io/temporal/workflow/queryTests/QueryCausingReplayWithLocalActivityInLastWFTTest.java
@@ -1,0 +1,92 @@
+/*
+ *  Copyright (C) 2020 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.workflow.queryTests;
+
+import static org.junit.Assert.assertEquals;
+
+import io.temporal.api.common.v1.WorkflowExecution;
+import io.temporal.client.WorkflowClient;
+import io.temporal.client.WorkflowStub;
+import io.temporal.testing.internal.SDKTestOptions;
+import io.temporal.testing.internal.SDKTestWorkflowRule;
+import io.temporal.workflow.Workflow;
+import io.temporal.workflow.shared.TestActivities.TestActivitiesImpl;
+import io.temporal.workflow.shared.TestActivities.VariousTestActivities;
+import io.temporal.workflow.shared.TestWorkflows;
+import org.junit.Rule;
+import org.junit.Test;
+
+// Covers a very specific scenario where
+// 1. the last completed workflow task has local activity markers
+// 2. workflow execution is not completed
+// 3. the workflow is evicted from the cache
+// 4. workflow gets a "direct (legacy) query" to execute
+// https://github.com/temporalio/sdk-java/issues/1190
+public class QueryCausingReplayWithLocalActivityInLastWFTTest {
+
+  private final TestActivitiesImpl activitiesImpl = new TestActivitiesImpl();
+
+  @Rule
+  public SDKTestWorkflowRule testWorkflowRule =
+      SDKTestWorkflowRule.newBuilder()
+          .setWorkflowTypes(TestLocalActivityAndQueryWorkflow.class)
+          .setActivityImplementations(activitiesImpl)
+          .build();
+
+  @Test
+  public void testLocalActivityAndQuery() {
+    TestWorkflows.TestWorkflowWithQuery workflowStub =
+        testWorkflowRule.newWorkflowStub(TestWorkflows.TestWorkflowWithQuery.class);
+    WorkflowExecution execution = WorkflowClient.start(workflowStub::execute);
+    // Don't do waitForOKQuery wait here, it changes the structure of history events in a way that
+    // doesn't reproduce the problem
+    testWorkflowRule.waitForTheEndOfWFT(execution);
+
+    testWorkflowRule.invalidateWorkflowCache();
+    assertEquals("updated", workflowStub.query());
+    assertEquals("done", WorkflowStub.fromTyped(workflowStub).getResult(String.class));
+
+    // local activity is expected to be triggered only once
+    activitiesImpl.assertInvocations("activity");
+  }
+
+  public static final class TestLocalActivityAndQueryWorkflow
+      implements TestWorkflows.TestWorkflowWithQuery {
+
+    String message = "initial";
+
+    @Override
+    public String execute() {
+      VariousTestActivities localActivities =
+          Workflow.newLocalActivityStub(
+              VariousTestActivities.class,
+              SDKTestOptions.newLocalActivityOptions().toBuilder().build());
+      localActivities.activity();
+      message = "updated";
+      Workflow.sleep(4_000);
+      return "done";
+    }
+
+    @Override
+    public String query() {
+      return message;
+    }
+  }
+}

--- a/temporal-test-server/src/main/java/io/temporal/internal/testservice/TestWorkflowStoreImpl.java
+++ b/temporal-test-server/src/main/java/io/temporal/internal/testservice/TestWorkflowStoreImpl.java
@@ -339,11 +339,10 @@ class TestWorkflowStoreImpl implements TestWorkflowStore {
         if (event.getEventType() == EventType.EVENT_TYPE_WORKFLOW_TASK_STARTED) {
           if (!iterator.hasNext()
               || iterator.peek().getEventType() == EventType.EVENT_TYPE_WORKFLOW_TASK_COMPLETED) {
-            previousStaredEventId = startedEventId;
-            startedEventId = event.getEventId();
+            previousStaredEventId = event.getEventId();
+            startedEventId = previousStaredEventId;
           }
         } else if (WorkflowExecutionUtils.isWorkflowExecutionClosedEvent(event)) {
-          previousStaredEventId = startedEventId;
           startedEventId = 0;
           if (iterator.hasNext()) {
             throw Status.INTERNAL
@@ -353,6 +352,8 @@ class TestWorkflowStoreImpl implements TestWorkflowStore {
         }
       }
       task.setPreviousStartedEventId(previousStaredEventId);
+      // it's not a real workflow task and the server sends 0 for startedEventId for such a workflow
+      // task
       task.setStartedEventId(startedEventId);
       if (taskQueue.getTaskQueueName().equals(task.getWorkflowExecutionTaskQueue().getName())) {
         history.addAllEvents(events);

--- a/temporal-testing/src/main/java/io/temporal/testing/internal/SDKTestWorkflowRule.java
+++ b/temporal-testing/src/main/java/io/temporal/testing/internal/SDKTestWorkflowRule.java
@@ -39,19 +39,20 @@ import io.temporal.client.WorkflowStub;
 import io.temporal.common.interceptors.WorkerInterceptor;
 import io.temporal.internal.common.DebugModeUtils;
 import io.temporal.internal.common.WorkflowExecutionHistory;
+import io.temporal.internal.replay.WorkflowExecutorCache;
 import io.temporal.serviceclient.WorkflowServiceStubs;
 import io.temporal.testing.TestWorkflowEnvironment;
 import io.temporal.testing.TestWorkflowRule;
-import io.temporal.worker.Worker;
-import io.temporal.worker.WorkerFactoryOptions;
-import io.temporal.worker.WorkerOptions;
-import io.temporal.worker.WorkflowImplementationOptions;
+import io.temporal.worker.*;
 import io.temporal.workflow.Functions;
 import java.io.File;
 import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
 import java.time.Duration;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.ListIterator;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ScheduledFuture;
@@ -75,6 +76,8 @@ public class SDKTestWorkflowRule implements TestRule {
   private static final Logger log = LoggerFactory.getLogger(SDKTestWorkflowRule.class);
 
   private static final long DEFAULT_TEST_TIMEOUT_SECONDS = 10;
+
+  private static final long BUSY_WAIT_SLEEP_MS = 100;
 
   public static final String NAMESPACE = "UnitTest";
   public static final String UUID_REGEXP =
@@ -183,6 +186,11 @@ public class SDKTestWorkflowRule implements TestRule {
     /** Global test timeout. Default is 10 seconds. */
     public Builder setTestTimeoutSeconds(long testTimeoutSeconds) {
       this.testTimeoutSeconds = testTimeoutSeconds;
+      return this;
+    }
+
+    public Builder setInitialTimeMillis(long initialTimeMillis) {
+      testWorkflowRuleBuilder.setInitialTimeMillis(initialTimeMillis);
       return this;
     }
 
@@ -299,6 +307,31 @@ public class SDKTestWorkflowRule implements TestRule {
     }
   }
 
+  /** Waits till the end of the workflow task if there is a workflow task in progress */
+  public void waitForTheEndOfWFT(WorkflowExecution execution) {
+    WorkflowExecutionHistory initialHistory = getExecutionHistory(execution);
+
+    HistoryEvent lastEvent = initialHistory.getLastEvent();
+    if (isWFTInProgress(lastEvent)) {
+      // wait for completion of a workflow task in progress
+      long startEventId = lastEvent.getEventId();
+      while (true) {
+        History history = getHistory(execution);
+        ListIterator<HistoryEvent> historyEventListIterator =
+            history.getEventsList().listIterator(history.getEventsList().size());
+        HistoryEvent previous;
+        for (previous = historyEventListIterator.previous();
+            historyEventListIterator.hasPrevious() && previous.getEventId() > startEventId;
+            previous = historyEventListIterator.previous()) {
+          if (!isWFTInProgress(historyEventListIterator.previous())) {
+            return;
+          }
+        }
+        busyWaitSleep();
+      }
+    }
+  }
+
   public WorkflowClient getWorkflowClient() {
     return testWorkflowRule.getWorkflowClient();
   }
@@ -341,13 +374,20 @@ public class SDKTestWorkflowRule implements TestRule {
   }
 
   /** Used to ensure that workflow first workflow task is executed. */
-  public static void waitForOKQuery(WorkflowStub stub) {
+  public static void waitForOKQuery(Object anyStub) {
+    WorkflowStub untypedStub;
+    if (anyStub instanceof WorkflowStub) {
+      untypedStub = (WorkflowStub) anyStub;
+    } else {
+      untypedStub = WorkflowStub.fromTyped(anyStub);
+    }
     while (true) {
       try {
-        String stackTrace = stub.query(QUERY_TYPE_STACK_TRACE, String.class);
+        String stackTrace = untypedStub.query(QUERY_TYPE_STACK_TRACE, String.class);
         if (!stackTrace.isEmpty()) {
           break;
         }
+        busyWaitSleep();
       } catch (WorkflowQueryException e) {
         // Ignore
       }
@@ -420,6 +460,44 @@ public class SDKTestWorkflowRule implements TestRule {
       }
     } else {
       testWorkflowRule.getTestEnvironment().sleep(d);
+    }
+  }
+
+  /** Causes eviction of all workflows in the worker cache */
+  // TODO replace the horrible reflection implementation with a normal protected access by hiding
+  //  WorkerFactory under an interface.
+  public void invalidateWorkflowCache() {
+    WorkerFactory workerFactory = testWorkflowRule.getTestEnvironment().getWorkerFactory();
+    try {
+      Method getCache = WorkerFactory.class.getDeclaredMethod("getCache");
+      getCache.setAccessible(true);
+      WorkflowExecutorCache cache = (WorkflowExecutorCache) getCache.invoke(workerFactory);
+      cache.invalidateAll();
+      while (cache.size() > 0) {
+        busyWaitSleep();
+      }
+    } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static boolean isWFTInProgress(HistoryEvent event) {
+    EventType eventType = event.getEventType();
+    switch (eventType) {
+      case EVENT_TYPE_WORKFLOW_TASK_STARTED:
+      case EVENT_TYPE_WORKFLOW_TASK_SCHEDULED:
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  private static void busyWaitSleep() {
+    try {
+      Thread.sleep(BUSY_WAIT_SLEEP_MS);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
     }
   }
 }

